### PR TITLE
[Internal] Add @retrieval in _experimental as an experimental feature.

### PR DESCRIPTION
**Pull request by thy09**
# Description

Add a simple @retrieval in experimental to help user define retrieval spans.

Samle code:
```python
def dummy_document(query):
    return {"content": "This is a dummy document.", "score": random.random()}
@retrieval
def my_search(query):
    return [dummy_document(query) for _ in range(3)]
```

Trace view:
image

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by lumoslnt — APPROVED** [review]

File: src/promptflow-tracing/promptflow/tracing/_span_enricher.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    _instance = None
+
+    def __init__(self):
+        self._type2enricher: Dict[str, SpanEnricher] = {}
+        self._base_enricher = SpanEnricher()
+
+    @classmethod
+    def get_instance(cls) -> "SpanEnricherManager":
+        if cls._instance is None:
+            cls._instance = SpanEnricherManager()
```
Comment (zzn2) [inline src/promptflow-tracing/promptflow/tracing/_span_enricher.py:26]: Should we add lock here?

File: src/promptflow-tracing/promptflow/tracing/_span_enricher.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+
+class SpanEnricher:
+    def __init__(self):
+        pass
+
+    def enrich(self, span, inputs, output):
+        from ._trace import enrich_span_with_output
+
+        enrich_span_with_output(span, output)
```
Comment (zzn2) [inline src/promptflow-tracing/promptflow/tracing/_span_enricher.py:13]: Why don't we enrich inputs here?

File: src/promptflow-tracing/promptflow/tracing/_retrieval.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+from opentelemetry.trace import Span
+
+from ._span_enricher import SpanEnricher, SpanEnricherManager
+from ._trace import TraceType, _traced
+
+keys = ["id", "score", "content", "metadata"]
+
+
+class RetrievalSpanEnricher(SpanEnricher):
```
Comment (zzn2) [inline src/promptflow-tracing/promptflow/tracing/_retrieval.py:12]: Consider separate into two PRs?
* One is for adding SpanEnricher and refactoring the existing spans.
* Another one is adding RetrievalSpan.

**Review by zzn2 — DISMISSED** [review]

File: src/promptflow-tracing/promptflow/tracing/_span_enricher.py (hunk 2 above)
Comment (thy09) [reply src/promptflow-tracing/promptflow/tracing/_span_enricher.py:13]: Will add a todo here, enrich inputs is before calling the function, this enrich is after calling the function.

File: src/promptflow-tracing/promptflow/tracing/_span_enricher.py (hunk 1 above)
Comment (thy09) [reply src/promptflow-tracing/promptflow/tracing/_span_enricher.py:26]: Actually such instance would be initialized while import the code, so currently we don't need to worry about this.

File: src/promptflow-tracing/promptflow/tracing/_retrieval.py (hunk 3 above)
Comment (thy09) [reply src/promptflow-tracing/promptflow/tracing/_retrieval.py:12]: Refactored.

**Review by zzn2 — APPROVED** [review]

**Pull request closed without merging**